### DISCUSSION TITLE
[Node Operator Question] Uneven archive node performance under load

### DISCUSSION THREAD

**Discussion by jkbrsn** [ethereum-optimism/developers#45] [Node Operators]
### Issue Description

Hello everyone,

I was referred here from the Optimism Discord after asking this question, so I'll pose it here as well:

We're running two Optimism archive nodes (legacy node + Bedrock/OP node) and have run into an elusive issue. One of our nodes is faring much worse under load than the other, becoming slower to respond and even falling behind in block height. The underlying hardware/infra is exactly the same for both nodes, and it is well above and minimum requirements listed in the docs. Both nodes are also running the same versions, `op-geth v1.101304.0` and `op-node v1.3.0`, at the moment, though the issue started to appear before they were upgraded to the latest version.

I've tried experimenting with the `op-geth` settings `--snapshot=... --cache=... --maxpeers=...` but they don't seem to have any effect. Particularly the `--snapshot` setting I thought would have at least some kind of impact, from what I understand that it does (I'm just a node runner though, actual blockchain client knowledge isn't very deep). On the affected node, the `--snapshot` setting was `true`, with state snapshot generated, when the issue appeared while it on the unaffected node was `false`.

Does this ring a bell for anyone? Any advice on what other configurations I might try out?

Cheers,
Jakob

**Comment by ghost** [comment]
They're archive nodes and have nothing to do with optimism actually, it is a op-geth issue.

You may try out some alternative EL nodes but notice that they are unstable, unproven to run for mission critical tasks.

https://github.com/testinprod-io/op-erigon
https://github.com/anton-rs/op-reth/tree/clabby/op-reth

**Reply by jkbrsn** [reply]
> They're archive nodes and have nothing to do with optimism actually, it is a geth issue.

So where would you advise to go looking for a solution, [official geth repo](https://github.com/ethereum/go-ethereum)? And is there any way I can tell if this is a `geth` issue or an `op-geth` issue?

Thanks for listing alternatives but we'll have to stick to what's proven for now.

**Reply by sbvegan** [reply]
`op-geth` is very close to `geth`; so generally speaking, their behaviors and fixes will be the same. You can checkout out their diffs here: https://op-geth.optimism.io/

Those alternative clients are awesome, but unstable at the moment.

**Reply by ghost** [reply]
Ah sorry for it, it could be the op-geth issue but if you encounter any RPC problem it will likely be inherited from upstream geth codebase

**Reply by jkbrsn** [reply]
> Those alternative clients are awesome, but unstable at the moment.

Totally agree, we've opted to use Erigon over Geth for some of our Ethereum nodes so would totally be down to use `op-erigon` once it's more proven.

**Comment by sbvegan** [comment]
Hey @jakobilobi, can you provide the following:

- system specs
- `op-node` configuration
- `op-geth` configuration

**Reply by jkbrsn** [reply]
Sure thing!

The node's system, running both `op-node` and `op-geth`, has 8 CPUs and 32GB RAM. The same goes for our other Optimism mainnet archive node, which doesn't show these issues.

`op-node` configuration (CLI command):
```
--l2.jwt-secret=<jwtsecret-path> --l2=<l2-address> --rpc.addr=0.0.0.0 --metrics.enabled --metrics.addr=0.0.0.0 --metrics.port=7300 --log.level=debug --l1=<l1-address> --rpc.port=9545 --network=mainnet
```

`op-geth` configuration (CLI command):
```
--ws --ws.addr=0.0.0.0 --ws.origins="*" --http --http.addr=0.0.0.0 --http.vhosts="*" --http.corsdomain="*" --authrpc.addr=localhost --authrpc.jwtsecret=<jwtsecret-path> --authrpc.port=8551 --authrpc.vhosts="*" --datadir=<datadir-path> --verbosity=3 --rollup.disabletxpoolgossip=true --nodiscover --syncmode=full --maxpeers=0 --metrics --rollup.sequencerhttp=<sequencer-address> --rollup.historicalrpc=<rpc-address> --gcmode=archive --networkid=10 --http.api=eth,rollup,net,web3,debug --http.port=8545 --ws.port=8546 --snapshot=false --cache=4096 --metrics.influxdb --metrics.influxdb.endpoint=<influxdb-address> --metrics.influxdb.username=<username> --metrics.influxdb.password=<password>
```

**Reply by jkbrsn** [reply]
I could add another piece of information as well; the node showing issues has about 7x the I/O read/write usage (40 MB/s vs 290 MB/s avg.) compared to the non-faulty node, even though they're supposedly running the exact same thing, on the same type of hardware. Perhaps that could be a clue to someone more knowledgeable.

**Reply by opfocus** [reply]
> even falling behind in block height.


Please note that I lack knowledge in this area, and I speculate that this might be the crux of the issue.

**Reply by sbvegan** [reply]
One thing you can try is setting `--cache=8192`, 8GB will help with performance here. We also want to make sure your `--snapshot=true` this our default recommendation right now.

> I could add another piece of information as well; the node showing issues has about 7x the I/O read/write usage (40 MB/s vs 290 MB/s avg.) compared to the non-faulty node

It would be helpful if we could see the cache / RAM usage, but this could also because `--snapshots=false` vs a recommended `--snapshots=true`

**Reply by jkbrsn** [reply]
Trying out these settings, will update here if I see changes in performance.

RAM usage has not been out of the ordinary, e.g. very similar to our other, functioning, node.

**Reply by sbvegan** [reply]
Hey @jakobilobi just checking back in, have you seen any improvements or any logs that might give us a clue to why your underperforming node is behaving that way?

**Reply by sbvegan** [reply]
Hey @jakobilobi, just following up one more time. Anything new to report?

**Reply by jkbrsn** [reply]
Hey, sorry for being ignorant of my notifications. Lots going on!

After doing some more debugging and having a look at our infrastructure we're quite confident at this point that the issue didn't have anything to do with the client software. Not as the main reason at least. I'll be closing this discussion, thanks for all the help and answers!

**Reply by sbvegan** [reply]
No worries, we're happy to help. Thanks for the follow up!

**Comment by opfocus** [comment]
btw, how do you monitor these status data? Have you started a dedicated monitoring program (this is another topic and is essentially irrelevant to the current problem)

**Reply by jkbrsn** [reply]
Irrelevant or not, happy to share :)

We export node client metrics via the supported settings (`--metrics.influxdb` etc.) and system metrics for the container running the client via our own setup, which uses Prometheus. The client metrics themselves don't have such low level info as disk I/O usage and similar, so that's sourced from the node monitoring.

**Reply by opfocus** [reply]
Thanks for share!

**Comment by jkbrsn** [comment]
Found that there was a more probable external reason for the issues we saw with the node, thanks for your help though!
